File: 0827-making-a-large-island/0827-making-a-large-island.cpp
class DisjointSet{
    public:
        vector<int> parent,size,rank;

        DisjointSet(int n){
            parent.resize(n+1);
            size.resize(n+1);
            rank.resize(n+1);

            for(int i=0;i<=n;i++){
                parent[i] = i;
                rank[i] = 0;
                size[i] = 1;
            }
        }

        int findParent(int node){
            if(node == parent[node]) return node;

            return parent[node] = findParent(parent[node]);
        }

        void unionByRank(int u,int v){
            int pu = findParent(u);
            int pv = findParent(v);

            if(pu == pv) return;

            if(rank[pu]< rank[pv]){
                parent[pu] = pv;
            }else if(rank[pu] > rank[pv]){
                parent[pv] = pu;
            }else{
                parent[pv] = pu;
                rank[pu]+=1;
            }
        }

        void unionBySize(int u,int v){
            int pu = findParent(u);
            int pv = findParent(v);

            if(pu == pv) return;

            if(size[pu] < size[pv]){
                parent[pu] = pv;
                size[pv] += size[pu];
            }else{
                parent[pv] = pu;
                size[pu] += size[pv];
            }
        }


};
class Solution {
private:
    bool isValid(int newr,int newc,int n){
        return newr>=0 && newr<n && newc>=0 && newc<n ;
    }
public:
    int largestIsland(vector<vector<int>>& grid) {
        
        int n = grid.size();
        DisjointSet ds(n*n);

        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j] == 1){
                    int dr[] = {-1,0,1,0};
                    int dc[] = {0,1,0,-1};

                    for(int k = 0;k<4;k++){
                        int newrow = i+dr[k];
                        int newcol = j + dc[k];

                        if(isValid(newrow,newcol,n) && grid[newrow][newcol] == 1){
                            int node = i*n + j;
                            int adjNode = newrow*n + newcol;

                            ds.unionBySize(node,adjNode);
                        }
                    }
                }
            }
        }

        int ans = 0;

        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j] == 1) continue;

                set<int> components;
                int currentSize = 1;

                int dr[] = {-1,0,1,0};
                int dc[] = {0,1,0,-1};
                int size = 1;

                for(int k = 0;k<4;k++){
                    int newr = i + dr[k];
                    int newc = j + dc[k];

                    if(isValid(newr,newc,n) && grid[newr][newc] == 1){
                        int adjNode = newr*n + newc;

                        int root = ds.findParent(adjNode);
                        components.insert(root);
                    }
                }

                for(auto it: components){
                    currentSize += ds.size[it];
                }
                ans = max(ans,currentSize);

            }
        }


        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                int node = i*n + j;

                ans = max(ans,ds.size[ds.findParent(node)]);
            }
        }

        return ans;


    }
};

File: 0827-making-a-large-island/0827-making-a-large-island.java
class DisjointSet{
    List<Integer> parent = new ArrayList<>();
    List<Integer> rank = new ArrayList<>();
    List<Integer> size = new ArrayList<>();
    public DisjointSet(int n){
        for(int i=0;i<=n;i++){
            parent.add(i);
            rank.add(0);
            size.add(1);
        }
    }

    int findParent(int node){
        if(parent.get(node) == node) return node;

        int ulp = findParent(parent.get(node));
        parent.set(node,ulp);

        return parent.get(node);
    }

    void unionByRank(int u,int v){
        int pu = findParent(u);
        int pv = findParent(v);

        if(pu == pv) return;

        if(rank.get(pu) < rank.get(pv)){
            parent.set(pu,pv);
        }else if(rank.get(pu) > rank.get(pv)){
            parent.set(pv,pu);
        }else{
            parent.set(pv,pu);
            rank.set(pu,rank.get(pu)+1);
        }
    }

    void unionBySize(int u,int v){
        int pu = findParent(u);
        int pv = findParent(v);

        if(pu == pv) return;

        if(size.get(pu) < size.get(pv)){
            parent.set(pu,pv);
            int sizeul = size.get(pu) + size.get(pv);
            size.set(pv,sizeul);
        }else{
            parent.set(pv,pu);
            int sizeul = size.get(pu) + size.get(pv);
            size.set(pu,sizeul);
        }
    }
}


class Solution {
    public int largestIsland(int[][] grid) {

        int n = grid.length;

        DisjointSet ds = new DisjointSet(n*n);

        for(int row = 0;row<n;row++){
            for(int col = 0;col<n;col++){
                if(grid[row][col] == 1){
                    int dr[] = {-1,0,1,0};
                    int dc[] = {0,1,0,-1};

                    for(int ind = 0;ind<4;ind++){
                        int newr = row + dr[ind];
                        int newc = col + dc[ind];

                        if(newr>=0 && newr<n && newc>=0 && newc<n && grid[newr][newc] == 1){
                            int node = row*n + col;
                            int adjNode = newr*n + newc;
                            ds.unionBySize(node,adjNode);
                        }
                    }
                }
            }
        }

        int ans = 0;
        for(int row = 0;row<n;row++){
            for(int col = 0;col<n;col++){
                if(grid[row][col] == 1) continue;

                Set<Integer> component = new HashSet<>();

                int dr[] = {-1,0,1,0};
                int dc[] = {0,1,0,-1};

                for(int ind = 0;ind<4;ind++){
                    int newr = row + dr[ind];
                    int newc = col + dc[ind];

                    if(newr>=0 && newr<n && newc>=0 && newc<n && grid[newr][newc] ==1){
                        int adjNode = newr*n + newc;

                        int root = ds.findParent(adjNode);
                        component.add(root);

                    }
                }

                int currentSize = 1;

                for(Integer it: component){
                    currentSize += ds.size.get(it);

                }
                ans = Math.max(ans,currentSize);
            }
        }

        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j] == 1){
                    int node = i*n + j;
                    ans = Math.max(ans,ds.size.get(ds.findParent(node)));
                }
            }
        }

        return ans;
    }
}